new/property.md

```diff
@@ -0,0 +1,206 @@
+
+```ruby
+file '/x.txt' do
+  content lazy { IO.read("/otherfile.txt") }
+end
+```
+
+### Validation
+
+There are a number of validation parameters to `property` that affect its behavior. If multiple of these are specified, they must *all* succeed.
+
+#### callbacks, kind_of, respond_to, cannot_be, regex, equal_to, required
+
+These function identically to the equivalent functionality on `attribute`.
+
+##### RSpec matchers
+
+`must` allows for rspec 3 matchers to be passed, and will validate them and print failures.
+
+```ruby
+include RSpec::Matchers
+property :path, String, must: start_with('/')
+```
+
+#### must_match
+
+This new option takes an array of values which use Ruby's universal matching
+operator, `===`.  This means that you can type this:
+
+```ruby
+property :x, must_match: [String, :a, :b, nil]
+```
+
+### Other options
+
+#### coerce
+
+`coerce` is a proc run when the user sets a non-lazy value, or reads a lazy or default value. It allows normalization of input, which makes it simple to create expressive interfaces while preserving a simple programming model that knows what to expect:
+
+```ruby
+class File < Chef::Resource
+  attribute :mode, coerce: proc { |v| mode.is_a?(String) ? mode.to_s(8) : mode }
+end
+```
+
+`coerce` procs are run in the context of the instance, so that they have access to other attributes and methods.
+
+#### default
+
+This works similarly to `attribute`, except that `lazy` values are automatically run in the context of the *instance*:
+
+```ruby
+class MysqlInstall < Chef::Resource
+  property :root_path, String, default: '/'
+  property :config_path, String, default: lazy { File.join(root_path, 'config') }
+end
+```
+
+#### name_attribute
+
+Same as before. Causes the attribute to be explicitly set to the name passed to the constructor. To wit:
+
+```ruby
+class MyResource < Chef::Resource
+  property :path, name_attribute: true
+end
+
+my_resource 'foo' do
+  name 'bar'
+  puts path #=> foo
+end
+```
+
+#### patchy
+
+Properties declare whether they are patchy (meaning resource actions will not change the on-disk value) or not patchy by specifying `patchy: true|false`. The primary effect of this is to prevent the property from being cloned during Chef's clone process (which happens when you declare a resource twice with different properties):
+
+
+```ruby
+file '/mystuff/x.txt' do
+  mode 0666
+end
+
+<long series of actions ...>
+
+execute 'chmod /mystuff 0777'
+
+<long series of actions ...>
+
+# If mode is declared `patchy: false`, we will change mode back to 0666 here.
+# If mode is declared `patchy: true`, we leave mode at 0777.
+file '/mystuff/x.txt' do
+  content 'Hello World'
+end
+```
+
+The reason being, a patchy property is one that *leaves the value alone* unless the user actually says they want to change it.  Cloning values from other resources violates that.
```

Comment:
When we drop resource cloning this just becomes a no-op though.   Is there a huge pressing use case for this? because I haven't seen any bugs about it.  We're going to force people to think about it, people will misuse it, then we'll drop it and carry around this no-op keyword in recipes going forwards, and inevitably write a 'consider dropping useless patchy modifier' foodcritic rule.  It really smells like something we should not implement only to take away later.